### DISCUSSION TITLE
Failed to see trace in Jaeger UI

### DISCUSSION THREAD

**Discussion by hongbo-miao** [linkerd/linkerd2#6434] [Q&A] [answered]
I am using Linkerd. I am trying to show trace in  Jaeger UI.

I have read these related issues, but none of them has accepted answers:
https://github.com/linkerd/linkerd2/discussions/5472
https://github.com/linkerd/linkerd2/discussions/5095
https://github.com/linkerd/linkerd2/discussions/4777

Following the Linkerd [tracing tutorial](https://linkerd.io/2.10/tasks/distributed-tracing/) which uses `emojivoto` app as demo, I am trying to repeat it in my own app.

### 1. emojivoto App

For the demo app **emojivoto**, after running

```shell
linkerd inject https://run.linkerd.io/emojivoto.yml | kubectl apply -f -
```

I can see `linkerd-proxy` got added to Jaeger UI service list first.

Then after set env by

```shell
kubectl -n emojivoto set env --all deploy OC_AGENT_HOST=collector.linkerd-jaeger:55678
```

I can see all server services `vote-bot`, `voting`, `web`, plus earlier `linkerd-proxy` got added to Jaeger UI service list.

![image](https://user-images.githubusercontent.com/3375461/124371355-a4f97280-dcb3-11eb-996c-f4ef9b09e880.png)

### 2. My app

Based on `emojivoto` tutorial. I am trying to add trace in [my project](https://github.com/Hongbo-Miao/hongbomiao.com) by basically coping emojivoto's codes.

My project has this structure:

- `web` is a frontend app.
- `api-go` is backend for frontend / a Go (Gin) server.
- `api-go-grpc` is a GRPC server.
- `opal-server` can be ignored because it is not related with this question.

I am using 

```yaml
annotations:
  linkerd.io/inject: enabled
```

for each my deployment yaml file such as this `api-go` [deployment yaml file](https://github.com/Hongbo-Miao/hongbomiao.com/blob/d53607edcbbad4034c0f14cdda73804e31742631/kubernetes/api-go-grpc-deployment.yaml#L15-L16) to use Automatic Proxy Injection.

After I start my app, all deployments successfully got meshed, and you can see my project traffic in the screenshot:
 
![image](https://user-images.githubusercontent.com/3375461/124371247-7929bd00-dcb2-11eb-97ec-170e4dead867.png)

Based on emojivoto's [commit that adding trace](https://github.com/BuoyantIO/emojivoto/commit/47a026c2e4085f4e536c2735f3ff3788b0870072), below are my codes (basically just copy and paste from emojivoto):

I have [this block of codes](https://github.com/Hongbo-Miao/hongbomiao.com/blob/d53607edcbbad4034c0f14cdda73804e31742631/api-go/cmd/grpc/main.go#L38-L53) in my GPRC server `api-go-grpc`:

```go
oce, err := ocagent.NewExporter(
	ocagent.WithInsecure(),
	ocagent.WithReconnectionPeriod(5*time.Second),
	ocagent.WithAddress(config.OpenCensusAgentHost+":"+config.OpenCensusAgentPort),  // this is "collector.linkerd-jaeger:55678" which also prints in my log
	ocagent.WithServiceName("api-go-grpc"))
if err != nil {
	log.Error().Err(err).Msg("ocagent.NewExporter")
}
trace.RegisterExporter(oce)

lis, err := net.Listen("tcp", ":"+config.GRPCPort)
if err != nil {
	log.Error().Err(err).Msg("net.Listen")
}

s := grpc.NewServer(grpc.StatsHandler(&ocgrpc.ServerHandler{}))
```

Also, I have [this block of codes](https://github.com/Hongbo-Miao/hongbomiao.com/blob/d53607edcbbad4034c0f14cdda73804e31742631/api-go/cmd/api/main.go#L24-L40) in my API Go (Gin) server `api-go`:

```go
oce, err := ocagent.NewExporter(
	ocagent.WithInsecure(),
	ocagent.WithReconnectionPeriod(5*time.Second),
	ocagent.WithAddress(config.OpenCensusAgentHost+":"+config.OpenCensusAgentPort),  // this is "collector.linkerd-jaeger:55678" which also prints in my log
	ocagent.WithServiceName("api-go"))
if err != nil {
	log.Error().Err(err).Msg("ocagent.NewExporter")
}
trace.RegisterExporter(oce)

r := routes.SetupRouter()
_ = http.ListenAndServe(
	":"+config.Port,
	&ochttp.Handler{
		Handler: r,
	},
)
```

Both OpenCensus Agent addresses are `collector.linkerd-jaeger:55678` which can be confirmed by my logs.

Also, the Linkerd screenshot below can double confirm my servers are pointing to `collector.linkerd-jaeger:55678` correctly:

![image](https://user-images.githubusercontent.com/3375461/124371618-2d791280-dcb6-11eb-847b-daad4bd79cf8.png)

However, if I only deploy my app (not together with `emojivoto` app), I am not be able to see my app's `linkerd-proxy` in the Jaeger UI service list, neither of my other servers. So no trace at all.

Any guide to help further debug will be appreciated. Thanks!

**Comment by cpretzer** [comment, marked as answer]
@Hongbo-Miao do you have an ingress configured? The Linkerd distributed tracing mechanism expects that a trace has been started, and the proxy will add the span ID to the existing trace ID.

If there is no trace initiated, then you won't see any spans connected together. That being said, I'd expect that the Jaeger UI would show several traces without span IDs.

**Reply by hongbo-miao** [reply]
Hi @cpretzer thanks for help! I assume you are referring https://kubernetes.github.io/ingress-nginx/user-guide/third-party-addons/opentracing/
I don't have Ingress set up. Also, my API server `api-go` and GPRC server `api-go-grpc` are both backends, so no need Ingress, right (?)

Currently, what I am trying to do is simply copy tracing codes from `emojivoto` app to make tracing work first. I am wondering how does `emojivoto` app start the trace because I didn't see [its yaml file](https://run.linkerd.io/emojivoto.yml) or [any step in the tutorial](https://linkerd.io/2.10/tasks/distributed-tracing/) involving Ingress neither.

**Reply by hongbo-miao** [reply]
Just add an update, I have added Ingress at https://github.com/Hongbo-Miao/hongbomiao.com/blob/main/kubernetes/ingress-service.yaml
All my Kubernetes files are at https://github.com/Hongbo-Miao/hongbomiao.com/tree/main/kubernetes

Ingress is for `web`.
- `web` is a frontend app.
- `api-go` is a server (has OpenCensus codes)
- `api-go-grpc` is a server (has OpenCensus codes)
- `opal-server` is a server (does not have OpenCensus codes).

![image](https://user-images.githubusercontent.com/3375461/124660572-f6cf1200-ded8-11eb-96d5-438aaaa18fc0.png)

However, still no luck. 😢

**Reply by cpretzer** [reply]
@Hongbo-Miao did you inject the ingress with the Linkerd proxy using the `linkerd.io/inject: ingress` annotation on the ingress  controller Deployment resource? I don't see it in the screenshot that you sent.

I have confirmed that the Ingress is expected to initiate the trace IDs, so you'll want to be sure that the Ingress pod is injected with the Linkerd proxy. This is the one-liner from the [Linkerd docs](https://linkerd.io/2.10/tasks/using-ingress/) for using ingress:

`kubectl get deployment <ingress-controller> -n <ingress-namespace> -o yaml | linkerd inject --ingress - | kubectl apply -f -`

**Reply by hongbo-miao** [reply]
Oh, thanks for the info! Wasn't aware of this `linkerd.io/inject: ingress` annotation. I will try it.
(I am still curious about why tracing works in the `emojivoto` project without Ingress)

**Reply by cpretzer** [reply]
@Hongbo-Miao I don't think I've seen the distributed tracing work in emojivoto without an Ingress

This is from the [distributed tracing docs](https://linkerd.io/2.10/features/distributed-tracing/#using-distributed-tracing):


> To use this feature, you’ll also need to introduce several additional components in your system., including an ingress layer that kicks off the trace on particular requests, a client library for your application (or a mechanism to propagate trace headers), a trace collector to collect span data and turn them into traces, and a trace backend to store the trace data and allow the user to view/query it.

And [this doc](https://linkerd.io/2.10/tasks/distributed-tracing/#recommendations) is relevant, too

> The ingress is an especially important component for distributed tracing because it creates the root span of each trace and is responsible for deciding if that trace should be sampled or not. Having the ingress make all sampling decisions ensures that either an entire trace is sampled or none of it is, and avoids creating “partial traces”.

**Reply by hongbo-miao** [reply]
Oh, the tracing in `emojivoto` demo definitely works without Ingress. Because I had a very [hard time](https://stackoverflow.com/questions/68274229/how-to-set-minikube-proxy-when-the-driver-is-hyperkit-or-virtualbox) adding Ingress in my current location, which is why I didn't try Ingress in the beginning. 😅

**Reply by hongbo-miao** [reply]
I injected by

```
kubectl get deployment ingress-nginx-controller --namespace=ingress-nginx --output=yaml | linkerd inject --ingress - | kubectl apply --filename=-
```

Then verified by 

```
kubectl describe pods/ingress-nginx-controller-7c56cf7f5-tctm4 --namespace=ingress-nginx | grep "linkerd.io/inject: ingress"
```

Ingress controller got meshed by Linkerd after injecting as the screenshot shows.

![image](https://user-images.githubusercontent.com/3375461/124673515-84b3f880-deeb-11eb-8a30-4bdfb5706c8d.png)

![image](https://user-images.githubusercontent.com/3375461/124673424-651cd000-deeb-11eb-9a10-6e4e455fd837.png)

However, my Jaeger UI still not show any tracing. Werid.
I also tried delete all mine under namespace `hm` and redeploy, but still same.

![image](https://user-images.githubusercontent.com/3375461/124673726-fd1ab980-deeb-11eb-82d5-3a4726ad6041.png)

**Reply by cpretzer** [reply]
> Oh, the tracing in `emojivoto` demo definitely works without Ingress. Because I had a very [hard time](https://stackoverflow.com/questions/68274229/how-to-set-minikube-proxy-when-the-driver-is-hyperkit-or-virtualbox) adding Ingress in my current location, which is why I didn't try Ingress in the beginning. 😅

You're right, the `vote-bot` code that generates load on the `web` service initiates the traceID: https://github.com/BuoyantIO/emojivoto/blob/main/emojivoto-web/cmd/vote-bot/main.go#L27

When using an ingress, the `vote-bot` service is not used, so the ingress should initiate the traces

**Reply by cpretzer** [reply]
> I injected by
> 
> ```
> kubectl get deployment ingress-nginx-controller --namespace=ingress-nginx --output=yaml | linkerd inject --ingress - | kubectl apply --filename=-
> ```
> 
> Then verified by
> 
> ```
> kubectl describe pods/ingress-nginx-controller-7c56cf7f5-tctm4 --namespace=ingress-nginx | grep "linkerd.io/inject: ingress"
> ```

It looks like the nginx ingress controller was properly injected with the Linkerd proxy. Did you follow the instructions to [enable opentracing](https://kubernetes.github.io/ingress-nginx/user-guide/third-party-addons/opentracing/#usage) in the ingress?

I don't see the annotation in the [ingress definition](https://github.com/Hongbo-Miao/hongbomiao.com/blob/main/kubernetes/hm-ingress.yaml#L7) in the repo that you linked.

**Reply by hongbo-miao** [reply]
Thanks for the [enable opentracing](https://kubernetes.github.io/ingress-nginx/user-guide/third-party-addons/opentracing/#usage) doc link! I wasn't aware of that before.

After ingress controller was properly injected with the Linkerd proxy. I patched my `ingress-nginx-controller` by

```shell
kubectl patch configmap ingress-nginx-controller --namespace=ingress-nginx --patch "$(cat ingress-nginx-controller-configmap-patch.yaml)"
```

**ingress-nginx-controller-configmap-patch.yaml**:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: ingress-nginx-controller
  namespace: ingress-nginx
data:
  enable-opentracing: "true"
  jaeger-collector-host: collector.linkerd-jaeger.svc.cluster.local
  # if I use "jaeger-collector-host: jaeger-agent.default.svc.cluster.local"
  # it will fail later when apply my Ingress yaml file
```

The [doc](https://kubernetes.github.io/ingress-nginx/user-guide/third-party-addons/opentracing/#usage) is using `jaeger-collector-host: jaeger-agent.default.svc.cluster.local`, if I use this I will get error when I apply my ingress file later by `kubectl apply --filename=hm-ingress.yaml`:

```
Error from server (BadRequest): error when creating "kubernetes/hm-ingress.yaml": admission webhook "validate.nginx.ingress.kubernetes.io" denied the request:
-------------------------------------------------------------------------------
Error: exit status 1
2021/07/07 15:24:01 [error] 3721#3721: Failed to construct tracer: Error resolving address: Name does not resolve
nginx: [error] Failed to construct tracer: Error resolving address: Name does not resolve
nginx: configuration file /tmp/nginx-cfg347946464 test failed
```

Here is my deployments:

![image](https://user-images.githubusercontent.com/3375461/124831299-9e1d7900-dfad-11eb-9f6c-4e03052d15f7.png)

So I updated to `jaeger-collector-host: collector.linkerd-jaeger.svc.cluster.local`. This time I can apply to my ingress yaml file `hm-ingress.yaml`, however, I still don't get any traces in Jaeger UI.

The [doc](https://kubernetes.github.io/ingress-nginx/user-guide/third-party-addons/opentracing/#usage) says

> NOTE: While the option is called jaeger-collector-host, you will need to point this to a jaeger-agent, and not the jaeger-collector component. Alternatively, you can set jaeger-endpoint and specify the full endpoint for uploading traces. This will use TCP and should be used for a collector rather than an agent.

Currently I am using `jaeger-collector-host: collector.linkerd-jaeger.svc.cluster.local` which is not an agent.
`linkerd-jaeger` currently only has a `collector`, no `agent`. Does that mean I need provide my own Jaeger agent? Thanks

(I will ask some of my coworkers who has done Jaeger and Ingress work  without service mesh for help too see if they can spot some errors)

**Reply by cpretzer** [reply]
@Hongbo-Miao, the Linkerd distributed tracing [docs](https://linkerd.io/2.10/tasks/distributed-tracing/#ingress) use the `zipkin-collector-host` setting, so I'd recommend using that, unless you're using an external Jaeger instance, that is to say if you decide to use something other than `linkerd-jaeger` in the future.

That being said, the `linkerd-jaeger` [chart](https://github.com/linkerd/linkerd2/blob/main/jaeger/charts/linkerd-jaeger/values.yaml#L74) deploys the [jaeger-all-in-one image](https://www.jaegertracing.io/docs/1.23/getting-started/#all-in-one) which deploys an agent listening on ports `5775`, `6831`,  and`6832`. I _think_ the port that you want to use is `6832` if you are going to use the jaeger agent and you will have to update the `Service` [resource](https://github.com/linkerd/linkerd2/blob/main/jaeger/charts/linkerd-jaeger/templates/tracing.yaml#L140) to expose that port.

I've never tried that myself, and since you're working through this PoC, I suggest sticking with the zipkin collector as outlined in the Linkerd docs.

**Reply by hongbo-miao** [reply]
Ah, it is using jaeger-all-in-one inside! Thanks for the info.
After changing to `zipkin-collector-host: collector.linkerd-jaeger`, now it works!
Not very intuitive though, as I thought it would be `jaeger-collector-host: collector.linkerd-jaeger`.

In future, I will change to external Jaeger instance and OpenTelemetry for further experiment. Will share back for how I make it. 😀

Thanks @cpretzer for support all time!

![image](https://user-images.githubusercontent.com/3375461/124931342-b4224c80-e034-11eb-9f6d-a33897e6cebf.png)

**Reply by cpretzer** [reply]
@Hongbo-Miao great to hear that you've got it working now.

The change to use OpenTelemetry is already in the edge releases thanks to [this PR](https://github.com/linkerd/linkerd2/pull/6188).

Regarding the `jaeger-collector-host` vs the `zipkin-collector-host`, please feel free to file a [docs issue](https://github.com/linkerd/website/issues/new) and PR to help us clarify the docs. 😄
